text/1021-rewrites.rst

```diff
@@ -0,0 +1,210 @@
+
+Note
+====
+
+This is a companion to
+`RFC 1020 <https://github.com/edgedb/rfcs/blob/master/text/1020-triggers.rst>`_,
+in that they work together to cover most of the desired use cases that motivated
+a previous `previous version triggers RFC <https://github.com/edgedb/rfcs/pull/70>`_
+
+Specification
+=============
+
+Synopsis::
+
+    type <type-name> "{"
+        { property | link } <name> -> <type> "{"
+            rewrite { insert | update } [, ... ] using (<expr>)
+        "}"
+    "}"
+
+We allow pointer definitions to be augmented with a ``rewrite`` rule.
+
+Expression interpretation
+-------------------------
+In ``<expr>``, ``__subject__`` (and the implicit path prefix) will be
+bound to the new value object being inserted or updated.
+
+In purely ``update`` rewrite rules, ``__old__`` will refer to the old
+version of the object.
+
+The special variable ``__fields__`` will be bound to all of the
+properties of ``<type-name>`` that were explicitly provided in the
+``insert`` or ``update``. This allows rewrite rules to distinguish
+between an explicitly provided empty set and no value being specified.
+
+(Whether ``<type-name>`` should refer to the object being mutated is
+kind of fraught; we leave that undecided for now. Some discussion of
+that
+`here
+<https://github.com/edgedb/edgedb/issues/4142#issuecomment-1386287450>_`.)
+
+Restrictions
+------------
+Each pointer can have one ``insert`` rewrite rule and one ``update``
+rewrite rule. They can be specified together if desired.
+
+Specifying a rewrite rule on a pointer is not allowed if an ancestor
+has already specified that kind of rewrite rule for the pointer.
+(Alternately: multiple rewrite rules are applied in MRO order?)
+
+
+Behavior
+--------
+
+When we perform ``insert`` or ``update``, after computing the
+values for the object, but before performing the operation or
+evaluating the write-side access policies, we evaluate each of the
+applicable rewrite rules and take the result as the actual value for
+the pointer.
+
+The rule for a pointer is applied whether or not that pointer is
+explicitly specified in the operation. ``__fields__`` can be used
+to determine whether it was and take action based on that.
+
+Note that for ``insert``, rewrite rules are *extremely* similar to
+``default`` (which can now refer to other pointers), but are applied
+even when a value was specified.
+
+
+Examples
+========
+
+Update a modification time field every time an object is modified::
+
+  type Entry {
+      # ...
+      required property mtime -> datetime {
+          rewrite insert, update using (datetime_of_statement())
+      }
+  };
+
+
+Update a modification time field every time an object is modified, but
+allow a manual override of it also::
+
+  type Entry {
+      # ...
+      required property mtime -> datetime {
+          rewrite insert, update using (
+              datetime_of_statement()
+              if 'mtime' not in __fields__
```

Comment:
What about a special name for "this pointer's name". `__field__` or some thing. It seems like it would be easy to rename the property here but miss the string in the `if` which would break this rewrite.

Reply:
It seems like it would be possible to write a linter that complains about the the pointer name being used in a string in a rewrite when we have that kind of tooling in the future.

Reply:
Yeah, that's a good point.

The other thing I considered was just having a boolean variable that indicates whether the current pointer was specified (`__specified__`, maybe? I'm sure there is some decent name option.). I went with suggesting `__fields__` because it seemed more general but maybe that generality is not super helpful?

Reply:
It seems like having all the modified names would enable more use cases like only updating `friend_count` when `friends` is updated?

Reply:
Another option is to invent a function-like intrinsic, e.g. `specified(.mtime)`